File: sdks/dart/lib/src/serializer.dart
import 'dart:convert';
import 'dart:typed_data';
import 'constants.dart';
import 'errors.dart';
import 'uleb128.dart';

/// BCS Serializer - Manual serialization API
class BcsSerializer {
  /// Create a serializer with optional initial capacity
  BcsSerializer([int initialCapacity = 256])
      : _buffer = Uint8List(initialCapacity);

  /// Internal buffer with dynamic growth
  Uint8List _buffer;
  int _size = 0;
  int _depth = 0;

  /// Scratch buffer for ULEB128 encoding
  final Uint8List _ulebBuffer = Uint8List(Uleb128.maxBytes);

  /// Ensure buffer has capacity for [needed] more bytes
  void _ensureCapacity(int needed) {
    // Check for overflow before addition
    if (needed > maxSequenceLength - _size) {
      throw BcsError.exceededMaxLength(_size + needed, maxSequenceLength);
    }
    final required = _size + needed;
    if (required <= _buffer.length) return;

    // Grow by doubling or to required size, whichever is larger
    // Guard against overflow in capacity calculation
    var newCapacity = _buffer.length * 2;
    if (newCapacity < 0 || newCapacity < required) newCapacity = required;
    final oldBuffer = _buffer;
    _buffer = Uint8List(newCapacity)..setRange(0, _size, oldBuffer);
  }

  // ==========================================================================
  // BOOLEAN
  // ==========================================================================

  /// Write a boolean value
  BcsSerializer writeBool(bool value) {
    _ensureCapacity(1);
    _buffer[_size++] = value ? 1 : 0;
    return this;
  }

  // ==========================================================================
  // UNSIGNED INTEGERS
  // ==========================================================================

  /// Write an unsigned 8-bit integer
  BcsSerializer writeU8(int value) {
    if (value < 0 || value > u8Max) {
      throw BcsError.integerOutOfRange('u8');
    }
    _ensureCapacity(1);
    _buffer[_size++] = value;
    return this;
  }

  /// Write an unsigned 16-bit integer (little-endian)
  BcsSerializer writeU16(int value) {
    if (value < 0 || value > u16Max) {
      throw BcsError.integerOutOfRange('u16');
    }
    _ensureCapacity(2);
    _buffer[_size] = value & 0xFF;
    _buffer[_size + 1] = (value >> 8) & 0xFF;
    _size += 2;
    return this;
  }

  /// Write an unsigned 32-bit integer (little-endian)
  BcsSerializer writeU32(int value) {
    if (value < 0 || value > u32Max) {
      throw BcsError.integerOutOfRange('u32');
    }
    _ensureCapacity(4);
    // Unrolled for performance
    _buffer[_size] = value & 0xFF;
    _buffer[_size + 1] = (value >> 8) & 0xFF;
    _buffer[_size + 2] = (value >> 16) & 0xFF;
    _buffer[_size + 3] = (value >> 24) & 0xFF;
    _size += 4;
    return this;
  }

  /// Write an unsigned 64-bit integer (little-endian)
  BcsSerializer writeU64(BigInt value) {
    if (value < BigInt.zero || value > u64Max) {
      throw BcsError.integerOutOfRange('u64');
    }
    _writeBigIntLE(value, 8);
    return this;
  }

  /// Write an unsigned 64-bit integer from int (little-endian).
  ///
  /// This is an optimized version of [writeU64] for values that fit in a
  /// native [int]. On the Dart VM, this supports the full u64 range for
  /// non-negative values. On Dart Web (JavaScript), [int] only has 53 bits
  /// of precision, so values larger than 2^53-1 may lose precision.
  ///
  /// For values that require full 64-bit precision on all platforms,
  /// use [writeU64] with [BigInt] instead.
  ///
  /// Throws [BcsError] if [value] is negative.
  BcsSerializer writeU64Int(int value) {
    if (value < 0) {
      throw BcsError.integerOutOfRange('u64');
    }
    _ensureCapacity(8);
    _buffer[_size] = value & 0xFF;
    _buffer[_size + 1] = (value >> 8) & 0xFF;
    _buffer[_size + 2] = (value >> 16) & 0xFF;
    _buffer[_size + 3] = (value >> 24) & 0xFF;
    _buffer[_size + 4] = (value >> 32) & 0xFF;
    _buffer[_size + 5] = (value >> 40) & 0xFF;
    _buffer[_size + 6] = (value >> 48) & 0xFF;
    _buffer[_size + 7] = (value >> 56) & 0xFF;
    _size += 8;
    return this;
  }

  /// Write an unsigned 128-bit integer (little-endian)
  BcsSerializer writeU128(BigInt value) {
    if (value < BigInt.zero || value > u128Max) {
      throw BcsError.integerOutOfRange('u128');
    }
    _writeBigIntLE(value, 16);
    return this;
  }

  /// Write an unsigned 256-bit integer (little-endian)
  BcsSerializer writeU256(BigInt value) {
    if (value < BigInt.zero || value > u256Max) {
      throw BcsError.integerOutOfRange('u256');
    }
    _writeBigIntLE(value, 32);
    return this;
  }

  // ==========================================================================
  // SIGNED INTEGERS
  // ==========================================================================

  /// Write a signed 8-bit integer
  BcsSerializer writeI8(int value) {
    if (value < i8Min || value > i8Max) {
      throw BcsError.integerOutOfRange('i8');
    }
    _ensureCapacity(1);
    _buffer[_size++] = value & 0xFF;
    return this;
  }

  /// Write a signed 16-bit integer (little-endian)
  BcsSerializer writeI16(int value) {
    if (value < i16Min || value > i16Max) {
      throw BcsError.integerOutOfRange('i16');
    }
    final unsigned = value & 0xFFFF;
    _ensureCapacity(2);
    _buffer[_size] = unsigned & 0xFF;
    _buffer[_size + 1] = (unsigned >> 8) & 0xFF;
    _size += 2;
    return this;
  }

  /// Write a signed 32-bit integer (little-endian)
  BcsSerializer writeI32(int value) {
    if (value < i32Min || value > i32Max) {
      throw BcsError.integerOutOfRange('i32');
    }
    final unsigned = value & 0xFFFFFFFF;
    _ensureCapacity(4);
    _buffer[_size] = unsigned & 0xFF;
    _buffer[_size + 1] = (unsigned >> 8) & 0xFF;
    _buffer[_size + 2] = (unsigned >> 16) & 0xFF;
    _buffer[_size + 3] = (unsigned >> 24) & 0xFF;
    _size += 4;
    return this;
  }

  /// Write a signed 64-bit integer (little-endian)
  BcsSerializer writeI64(BigInt value) {
    if (value < i64Min || value > i64Max) {
      throw BcsError.integerOutOfRange('i64');
    }
    _writeSignedBigIntLE(value, 8);
    return this;
  }

  /// Write a signed 128-bit integer (little-endian)
  BcsSerializer writeI128(BigInt value) {
    if (value < i128Min || value > i128Max) {
      throw BcsError.integerOutOfRange('i128');
    }
    _writeSignedBigIntLE(value, 16);
    return this;
  }

  /// Write a signed 256-bit integer (little-endian)
  BcsSerializer writeI256(BigInt value) {
    if (value < i256Min || value > i256Max) {
      throw BcsError.integerOutOfRange('i256');
    }
    _writeSignedBigIntLE(value, 32);
    return this;
  }

  // ==========================================================================
  // ULEB128
  // ==========================================================================

  /// Write a ULEB128-encoded length
  BcsSerializer writeUleb128(int value) {
    // Fast path for small values (0-127)
    if (value < 0x80) {
      _ensureCapacity(1);
      _buffer[_size++] = value;
      return this;
    }
    // Use scratch buffer for larger values
    final len = Uleb128.encodeInto(value, _ulebBuffer);
    _ensureCapacity(len);
    _buffer.setRange(_size, _size + len, _ulebBuffer);
    _size += len;
    return this;
  }

  // ==========================================================================
  // BYTES AND STRINGS
  // ==========================================================================

  /// Write fixed-length bytes (without length prefix)
  BcsSerializer writeFixedBytes(Uint8List data) {
    final len = data.length;
    _ensureCapacity(len);
    _buffer.setRange(_size, _size + len, data);
    _size += len;
    return this;
  }

  /// Write bytes with ULEB128 length prefix
  BcsSerializer writeBytes(Uint8List data) {
    _checkSequenceLength(data.length);
    // Pre-calculate total size needed
    final ulebSize = Uleb128.encodedSize(data.length);
    _ensureCapacity(ulebSize + data.length);
    writeUleb128(data.length);
    _buffer.setRange(_size, _size + data.length, data);
    _size += data.length;
    return this;
  }

  /// Write a UTF-8 string with ULEB128 length prefix
  BcsSerializer writeString(String value) {
    final bytes = utf8.encode(value);
    _checkSequenceLength(bytes.length);
    final ulebSize = Uleb128.encodedSize(bytes.length);
    _ensureCapacity(ulebSize + bytes.length);
    writeUleb128(bytes.length);
    _buffer.setRange(_size, _size + bytes.length, bytes);
    _size += bytes.length;
    return this;
  }

  // ==========================================================================
  // COMPOSITE TYPES
  // ==========================================================================

  /// Write an optional value
  BcsSerializer writeOption<T>(
    T? value,
    void Function(BcsSerializer, T) serializer,
  ) {
    if (value == null) {
      _ensureCapacity(1);
      _buffer[_size++] = 0;
    } else {
      _ensureCapacity(1);
      _buffer[_size++] = 1;
      serializer(this, value);
    }
    return this;
  }

  /// Write a vector with element serializer
  BcsSerializer writeVector<T>(
    List<T> values,
    void Function(BcsSerializer, T) serializer,
  ) {
    _checkSequenceLength(values.length);
    writeUleb128(values.length);
    for (final value in values) {
      serializer(this, value);
    }
    return this;
  }

  /// Write a map with key/value serializers (sorted by serialized key bytes)
  BcsSerializer writeMap<K, V>(
    Map<K, V> map,
    void Function(BcsSerializer, K) keySerializer,
    void Function(BcsSerializer, V) valueSerializer,
  ) {
    _checkSequenceLength(map.length);

    // Serialize all keys to get their byte representation
    // Reuse a single serializer for efficiency
    final entries = <(Uint8List, K, V)>[];
    final keySer = BcsSerializer(64); // Small initial capacity for keys
    for (final entry in map.entries) {
      keySer.clear();
      keySerializer(keySer, entry.key);
      // Must copy since we reuse the serializer
      entries.add((keySer.toBytesCopy(), entry.key, entry.value));
    }

    // Sort by key bytes (lexicographic)
    entries.sort((a, b) => _compareBytes(a.$1, b.$1));

    // Write length and entries
    writeUleb128(entries.length);
    for (final (keyBytes, _, value) in entries) {
      writeFixedBytes(keyBytes);
      valueSerializer(this, value);
    }

    return this;
  }

  /// Write an enum variant index (ULEB128)
  BcsSerializer writeVariantIndex(int index) {
    _enterContainer();
    writeUleb128(index);
    return this;
  }

  // ==========================================================================
  // CONTAINER DEPTH
  // ==========================================================================

  /// Enter a struct container for depth tracking
  BcsSerializer enterStruct([String name = '']) {
    _enterContainer();
    return this;
  }

  /// Leave the current struct container
  BcsSerializer leaveStruct() {
    _leaveContainer();
    return this;
  }

  /// Enter an enum container and write variant index
  BcsSerializer enterEnum(int index) {
    return writeVariantIndex(index);
  }

  /// Leave the current enum container
  BcsSerializer leaveEnum() {
    _leaveContainer();
    return this;
  }

  // ==========================================================================
  // OUTPUT
  // ==========================================================================

  /// Get the serialized bytes as a copy.
  ///
  /// Returns a new [Uint8List] containing a copy of the serialized data.
  /// This is safe to use even if the serializer is reused via [clear].
  ///
  /// For performance-critical code where you won't reuse the serializer,
  /// consider using [toBytesView] instead.
  Uint8List toBytes() {
    return Uint8List.fromList(_buffer.sublist(0, _size));
  }

  /// Get the serialized bytes as a view (zero-copy).
  ///
  /// **WARNING:** The returned [Uint8List] is a view into the internal buffer.
  /// If you call [clear] and reuse this serializer, the previously returned
  /// view will be corrupted. Only use this method when you're done with the
  /// serializer or need maximum performance.
  Uint8List toBytesView() {
    return Uint8List.sublistView(_buffer, 0, _size);
  }

  /// Alias for [toBytes] - returns a copy.
  Uint8List toBytesCopy() => toBytes();

  /// Get the current size of the buffer
  int get size => _size;

  /// Clear the buffer for reuse.
  ///
  /// **Note:** Any views returned by [toBytesView] will be invalidated.
  /// Use [toBytes] or [toBytesCopy] if you need to keep the data.
  void clear() {
    _size = 0;
    _depth = 0;
  }

  // ==========================================================================
  // PRIVATE HELPERS
  // ==========================================================================

  void _checkSequenceLength(int length) {
    if (length > maxSequenceLength) {
      throw BcsError.exceededMaxLength(length, maxSequenceLength);
    }
  }

  void _enterContainer() {
    _depth++;
    if (_depth > maxContainerDepth) {
      throw BcsError.exceededContainerDepth(_depth, maxContainerDepth);
    }
  }

  void _leaveContainer() {
    if (_depth > 0) {
      _depth--;
    }
  }

  /// Optimized BigInt write using direct byte extraction
  void _writeBigIntLE(BigInt value, int byteLength) {
    _ensureCapacity(byteLength);
    final mask = BigInt.from(0xFF);
    var v = value;
    for (var i = 0; i < byteLength; i++) {
      _buffer[_size + i] = (v & mask).toInt();
      v >>= 8;
    }
    _size += byteLength;
  }

  void _writeSignedBigIntLE(BigInt value, int byteLength) {
    final mask = (BigInt.one << (byteLength * 8)) - BigInt.one;
    final unsigned = value & mask;
    _writeBigIntLE(unsigned, byteLength);
  }

  static int _compareBytes(Uint8List a, Uint8List b) {
    final minLen = a.length < b.length ? a.length : b.length;
    for (var i = 0; i < minLen; i++) {
      if (a[i] < b[i]) return -1;
      if (a[i] > b[i]) return 1;
    }
    return a.length.compareTo(b.length);
  }
}

File: sdks/typescript/src/serializer.ts
/**
 * BCS Serializer - Manual serialization API.
 *
 * Provides explicit methods for serializing each BCS type.
 * Use this for full control over serialization.
 *
 * @example
 * ```ts
 * const ser = new BcsSerializer();
 * ser.writeU8(1);
 * ser.writeU64(100n);
 * ser.writeString("hello");
 * const bytes = ser.toBytes();
 * ```
 */

import { BcsError } from "./errors";
import * as uleb128 from "./uleb128";

// Constants
export const MAX_SEQUENCE_LENGTH = 2147483647; // 2^31 - 1
export const MAX_CONTAINER_DEPTH = 500;

// Integer bounds
const U8_MAX = 0xff;
const U16_MAX = 0xffff;
const U32_MAX = 0xffffffff;
const U64_MAX = 0xffffffffffffffffn;
const U128_MAX = (1n << 128n) - 1n;
const U256_MAX = (1n << 256n) - 1n;

const I8_MIN = -128;
const I8_MAX = 127;
const I16_MIN = -32768;
const I16_MAX = 32767;
const I32_MIN = -2147483648;
const I32_MAX = 2147483647;
const I64_MIN = -(1n << 63n);
const I64_MAX = (1n << 63n) - 1n;
const I128_MIN = -(1n << 127n);
const I128_MAX = (1n << 127n) - 1n;
const I256_MIN = -(1n << 255n);
const I256_MAX = (1n << 255n) - 1n;

// Cached TextEncoder for string serialization
const textEncoder = new TextEncoder();

// Initial buffer size
const INITIAL_CAPACITY = 256;

/**
 * BCS Serializer class.
 */
export class BcsSerializer {
  private buffer: Uint8Array;
  private size: number = 0;
  private depth: number = 0;

  constructor(initialCapacity: number = INITIAL_CAPACITY) {
    this.buffer = new Uint8Array(initialCapacity);
  }

  /** Ensure buffer has capacity for additional bytes */
  private ensureCapacity(needed: number): void {
    const required = this.size + needed;
    if (required <= this.buffer.length) return;

    // Grow by doubling or to required size
    let newCapacity = this.buffer.length * 2;
    if (newCapacity < required) newCapacity = required;
    const newBuffer = new Uint8Array(newCapacity);
    newBuffer.set(this.buffer.subarray(0, this.size));
    this.buffer = newBuffer;
  }

  // ==========================================================================
  // BOOLEAN
  // ==========================================================================

  /**
   * Serialize a boolean value.
   */
  writeBool(value: boolean): this {
    this.ensureCapacity(1);
    this.buffer[this.size++] = value ? 1 : 0;
    return this;
  }

  // ==========================================================================
  // UNSIGNED INTEGERS
  // ==========================================================================

  /**
   * Serialize an unsigned 8-bit integer.
   */
  writeU8(value: number): this {
    if (!Number.isInteger(value) || value < 0 || value > U8_MAX) {
      throw BcsError.valueOutOfRange("u8", value);
    }
    this.ensureCapacity(1);
    this.buffer[this.size++] = value;
    return this;
  }

  /**
   * Serialize an unsigned 16-bit integer (little-endian).
   */
  writeU16(value: number): this {
    if (!Number.isInteger(value) || value < 0 || value > U16_MAX) {
      throw BcsError.valueOutOfRange("u16", value);
    }
    this.ensureCapacity(2);
    this.buffer[this.size] = value & 0xff;
    this.buffer[this.size + 1] = (value >> 8) & 0xff;
    this.size += 2;
    return this;
  }

  /**
   * Serialize an unsigned 32-bit integer (little-endian).
   */
  writeU32(value: number): this {
    if (!Number.isInteger(value) || value < 0 || value > U32_MAX) {
      throw BcsError.valueOutOfRange("u32", value);
    }
    this.ensureCapacity(4);
    // Unrolled for performance
    this.buffer[this.size] = value & 0xff;
    this.buffer[this.size + 1] = (value >> 8) & 0xff;
    this.buffer[this.size + 2] = (value >> 16) & 0xff;
    this.buffer[this.size + 3] = (value >>> 24) & 0xff;
    this.size += 4;
    return this;
  }

  /**
   * Serialize an unsigned 64-bit integer (little-endian).
   */
  writeU64(value: bigint): this {
    if (value < 0n || value > U64_MAX) {
      throw BcsError.valueOutOfRange("u64", value);
    }
    this.writeBigIntLE(value, 8);
    return this;
  }

  /**
   * Serialize an unsigned 128-bit integer (little-endian).
   */
  writeU128(value: bigint): this {
    if (value < 0n || value > U128_MAX) {
      throw BcsError.valueOutOfRange("u128", value);
    }
    this.writeBigIntLE(value, 16);
    return this;
  }

  /**
   * Serialize an unsigned 256-bit integer (little-endian).
   */
  writeU256(value: bigint): this {
    if (value < 0n || value > U256_MAX) {
      throw BcsError.valueOutOfRange("u256", value);
    }
    this.writeBigIntLE(value, 32);
    return this;
  }

  // ==========================================================================
  // SIGNED INTEGERS
  // ==========================================================================

  /**
   * Serialize a signed 8-bit integer (two's complement).
   */
  writeI8(value: number): this {
    if (!Number.isInteger(value) || value < I8_MIN || value > I8_MAX) {
      throw BcsError.valueOutOfRange("i8", value);
    }
    this.ensureCapacity(1);
    this.buffer[this.size++] = value < 0 ? value + 256 : value;
    return this;
  }

  /**
   * Serialize a signed 16-bit integer (two's complement, little-endian).
   */
  writeI16(value: number): this {
    if (!Number.isInteger(value) || value < I16_MIN || value > I16_MAX) {
      throw BcsError.valueOutOfRange("i16", value);
    }
    const unsigned = value < 0 ? value + 0x10000 : value;
    this.ensureCapacity(2);
    this.buffer[this.size] = unsigned & 0xff;
    this.buffer[this.size + 1] = (unsigned >> 8) & 0xff;
    this.size += 2;
    return this;
  }

  /**
   * Serialize a signed 32-bit integer (two's complement, little-endian).
   */
  writeI32(value: number): this {
    if (!Number.isInteger(value) || value < I32_MIN || value > I32_MAX) {
      throw BcsError.valueOutOfRange("i32", value);
    }
    const unsigned = value < 0 ? value + 0x100000000 : value;
    this.ensureCapacity(4);
    this.buffer[this.size] = unsigned & 0xff;
    this.buffer[this.size + 1] = (unsigned >> 8) & 0xff;
    this.buffer[this.size + 2] = (unsigned >> 16) & 0xff;
    this.buffer[this.size + 3] = (unsigned >>> 24) & 0xff;
    this.size += 4;
    return this;
  }

  /**
   * Serialize a signed 64-bit integer (two's complement, little-endian).
   */
  writeI64(value: bigint): this {
    if (value < I64_MIN || value > I64_MAX) {
      throw BcsError.valueOutOfRange("i64", value);
    }
    const unsigned = value < 0n ? value + (1n << 64n) : value;
    this.writeBigIntLE(unsigned, 8);
    return this;
  }

  /**
   * Serialize a signed 128-bit integer (two's complement, little-endian).
   */
  writeI128(value: bigint): this {
    if (value < I128_MIN || value > I128_MAX) {
      throw BcsError.valueOutOfRange("i128", value);
    }
    const unsigned = value < 0n ? value + (1n << 128n) : value;
    this.writeBigIntLE(unsigned, 16);
    return this;
  }

  /**
   * Serialize a signed 256-bit integer (two's complement, little-endian).
   */
  writeI256(value: bigint): this {
    if (value < I256_MIN || value > I256_MAX) {
      throw BcsError.valueOutOfRange("i256", value);
    }
    const unsigned = value < 0n ? value + (1n << 256n) : value;
    this.writeBigIntLE(unsigned, 32);
    return this;
  }

  // ==========================================================================
  // ULEB128
  // ==========================================================================

  /**
   * Serialize a ULEB128-encoded unsigned integer.
   */
  writeUleb128(value: number): this {
    // Fast path for single byte (0-127)
    if (value < 0x80) {
      this.ensureCapacity(1);
      this.buffer[this.size++] = value;
      return this;
    }
    // Multi-byte path
    this.ensureCapacity(5);
    const len = uleb128.encodeInto(value, this.buffer, this.size);
    this.size += len;
    return this;
  }

  // ==========================================================================
  // BYTES AND STRINGS
  // ==========================================================================

  /**
   * Serialize a byte array (length-prefixed with ULEB128).
   */
  writeBytes(value: Uint8Array): this {
    if (value.length > MAX_SEQUENCE_LENGTH) {
      throw BcsError.exceededMaxLength(value.length);
    }
    // Pre-allocate for length + data
    const ulebSize = uleb128.encodedSize(value.length);
    this.ensureCapacity(ulebSize + value.length);
    this.writeUleb128(value.length);
    this.buffer.set(value, this.size);
    this.size += value.length;
    return this;
  }

  /**
   * Serialize a UTF-8 string (length-prefixed with ULEB128).
   */
  writeString(value: string): this {
    const bytes = textEncoder.encode(value);
    return this.writeBytes(bytes);
  }

  /**
   * Serialize fixed-length bytes (no length prefix).
   */
  writeFixedBytes(value: Uint8Array, length: number): this {
    if (value.length !== length) {
      throw new Error(`Expected ${length} bytes, got ${value.length}`);
    }
    this.ensureCapacity(length);
    this.buffer.set(value, this.size);
    this.size += length;
    return this;
  }

  // ==========================================================================
  // OPTION
  // ==========================================================================

  /**
   * Serialize an optional value.
   *
   * @param value - The value or null/undefined
   * @param serializer - Function to serialize the inner value
   */
  writeOption<T>(
    value: T | null | undefined,
    serializer: (ser: BcsSerializer, v: T) => void
  ): this {
    if (value === null || value === undefined) {
      this.ensureCapacity(1);
      this.buffer[this.size++] = 0;
    } else {
      this.ensureCapacity(1);
      this.buffer[this.size++] = 1;
      serializer(this, value);
    }
    return this;
  }

  // ==========================================================================
  // VECTOR
  // ==========================================================================

  /**
   * Serialize a vector of values.
   *
   * @param values - The array of values
   * @param serializer - Function to serialize each element
   */
  writeVector<T>(values: T[], serializer: (ser: BcsSerializer, v: T) => void): this {
    if (values.length > MAX_SEQUENCE_LENGTH) {
      throw BcsError.exceededMaxLength(values.length);
    }
    this.writeUleb128(values.length);
    for (const value of values) {
      serializer(this, value);
    }
    return this;
  }

  // ==========================================================================
  // ENUM
  // ==========================================================================

  /**
   * Write an enum variant index (ULEB128).
   */
  writeVariantIndex(index: number): this {
    return this.writeUleb128(index);
  }

  // ==========================================================================
  // CONTAINER DEPTH
  // ==========================================================================

  /**
   * Enter a struct container for depth tracking.
   * @throws BcsError if container depth exceeds MAX_CONTAINER_DEPTH (500)
   */
  enterStruct(): this {
    this.enterContainer("struct");
    return this;
  }

  /**
   * Leave a struct container.
   */
  leaveStruct(): this {
    this.leaveContainer();
    return this;
  }

  /**
   * Enter an enum container for depth tracking and write variant index.
   * @param variantIndex - The enum variant index
   * @throws BcsError if container depth exceeds MAX_CONTAINER_DEPTH (500)
   */
  enterEnum(variantIndex: number): this {
    this.enterContainer("enum");
    this.writeVariantIndex(variantIndex);
    return this;
  }

  /**
   * Leave an enum container.
   */
  leaveEnum(): this {
    this.leaveContainer();
    return this;
  }

  /**
   * Get the current container depth.
   */
  get containerDepth(): number {
    return this.depth;
  }

  private enterContainer(name: string): void {
    if (this.depth >= MAX_CONTAINER_DEPTH) {
      throw BcsError.exceededContainerDepth(name);
    }
    this.depth++;
  }

  private leaveContainer(): void {
    if (this.depth > 0) {
      this.depth--;
    }
  }

  // ==========================================================================
  // MAP
  // ==========================================================================

  /**
   * Serialize a map (sorted by key bytes).
   *
   * @param entries - The map entries as an array of [key, value] pairs or Map
   * @param keySerializer - Function to serialize keys
   * @param valueSerializer - Function to serialize values
   */
  writeMap<K, V>(
    entries: Map<K, V> | [K, V][],
    keySerializer: (ser: BcsSerializer, k: K) => void,
    valueSerializer: (ser: BcsSerializer, v: V) => void
  ): this {
    const entryArray = entries instanceof Map ? Array.from(entries.entries()) : entries;

    if (entryArray.length > MAX_SEQUENCE_LENGTH) {
      throw BcsError.exceededMaxLength(entryArray.length);
    }

    // Serialize keys to get bytes for sorting
    const keyBytesEntries = entryArray.map(([key, value]) => {
      const keySer = new BcsSerializer();
      keySerializer(keySer, key);
      return { keyBytes: keySer.toBytes(), key, value };
    });

    // Sort by key bytes
    keyBytesEntries.sort((a, b) => compareBytes(a.keyBytes, b.keyBytes));

    // Write length and sorted entries
    this.writeUleb128(keyBytesEntries.length);
    for (const { keyBytes, value } of keyBytesEntries) {
      this.ensureCapacity(keyBytes.length);
      for (const byte of keyBytes) {
        this.buffer[this.size++] = byte;
      }
      valueSerializer(this, value);
    }

    return this;
  }

  // ==========================================================================
  // UTILITY
  // ==========================================================================

  /**
   * Get the serialized bytes.
   */
  toBytes(): Uint8Array {
    return this.buffer.slice(0, this.size);
  }

  /**
   * Get the current length in bytes.
   */
  get length(): number {
    return this.size;
  }

  // ==========================================================================
  // PRIVATE HELPERS
  // ==========================================================================

  private writeBigIntLE(value: bigint, byteLength: number): void {
    this.ensureCapacity(byteLength);
    let v = value;
    for (let i = 0; i < byteLength; i++) {
      this.buffer[this.size + i] = Number(v & 0xffn);
      v >>= 8n;
    }
    this.size += byteLength;
  }
}

/**
 * Compare two byte arrays lexicographically.
 */
function compareBytes(a: Uint8Array, b: Uint8Array): number {
  const minLen = Math.min(a.length, b.length);
  for (let i = 0; i < minLen; i++) {
    if (a[i] !== b[i]) {
      return a[i] - b[i];
    }
  }
  return a.length - b.length;
}
